File: C-sharp/Labwork 5/ActivityScheduler.cs
using Labwork_5.MainFlow;

namespace Labwork_5
{
    public class ActivityScheduler
    {
        private static List<Event> s_activities = new List<Event>();

        /// <exception cref="ArgumentOutOfRangeException"></exception>
        public static void AddActivity(Event activity)
        {
            Validator.ValidateEventTime(activity);
            s_activities.Add(activity);
        }

        public static List<Event> GetActivitiesList()
        {
            return s_activities;
        }

        public static Meeting GetLatestMeeting()
        {
            return s_activities.Where(activity => activity is Meeting).MaxBy(activity => activity.DateTime) as Meeting;
        }

        public static void AssignDateToEvents(DateOnly date)
        {
            foreach (Event activity in ActivityScheduler.GetActivitiesList())
            {
                if (activity is Meeting meeting)
                {
                    meeting.DateTime = date.ToDateTime(TimeOnly.FromDateTime(meeting.DateTime));
                }
                else if (activity is Birthday birthday)
                {
                    birthday.DateTime = date.ToDateTime(TimeOnly.FromDateTime(birthday.DateTime));
                }
            }
        }
    }
}

File: Python/Labwork 5/activity_scheduler.py
from validator import *
from datetime import *
from events import *
import functions


class ActivityScheduler:
    __activities = []
    
    @staticmethod
    def add_activity(activity):
        functions.validate_event_time(activity, ActivityScheduler.get_activities_list())
        ActivityScheduler.__activities.append(activity)

    @staticmethod
    def get_activities_list():
        return ActivityScheduler.__activities

    @staticmethod
    def get_latest_meeting():
        meetings = list(filter(lambda activity: isinstance(activity, Meeting), ActivityScheduler.__activities))
        result = meetings[0]

        for meeting in meetings:
            if meeting.get_date_time() > result.get_date_time():
                result = meeting

        return result

    @staticmethod
    def assign_date_to_events(date):
        for activity in ActivityScheduler.get_activities_list():
            if isinstance(activity, Meeting):
                activity.set_date_time(datetime.combine(date, activity.get_date_time().time()))
            elif isinstance(activity, Birthday):
                activity.set_date_time(datetime.combine(date, activity.get_date_time().time()))
            